clippy_lints/src/methods/map_flatten.rs

```diff
@@ -1,83 +1,119 @@
+            format!("{}({})", method_to_use, closure_snippet),
+            applicability,
         );
     }
+}
 
-    // lint if caller of `.map().flatten()` is an Option or Result
-    let caller_type = match cx.typeck_results().expr_ty(recv).kind() {
-        ty::Adt(adt, _) => {
+fn try_get_caller_ty_name_and_method_name(
+    cx: &LateContext<'_>,
+    expr: &Expr<'_>,
+    caller_expr: &Expr<'_>,
+    map_arg: &Expr<'_>,
+) -> Option<(String, String)> {
+    if is_trait_method(cx, expr, sym::Iterator) {
+        if is_map_to_option(cx, map_arg) {
+            // `(...).map(...)` has type `impl Iterator<Item=Option<...>>
+            Some(("Iterator".to_string(), "filter_map".to_string()))
+        } else {
+            // `(...).map(...)` has type `impl Iterator<Item=impl Iterator<...>>
+            Some(("Iterator".to_string(), "flat_map".to_string()))
+        }
+    } else {
+        if let ty::Adt(adt, _) = cx.typeck_results().expr_ty(caller_expr).kind() {
             if cx.tcx.is_diagnostic_item(sym::Option, adt.did()) {
-                "Option"
+                return Some(("Option".to_string(), "and_then".to_string()));
             } else if cx.tcx.is_diagnostic_item(sym::Result, adt.did()) {
-                "Result"
-            } else {
-                return;
+                return Some(("Result".to_string(), "and_then".to_string()));
             }
+        }
+        None
+    }
+}
+
+fn is_map_to_option(cx: &LateContext<'_>, map_arg: &Expr<'_>) -> bool {
+    let map_closure_ty = cx.typeck_results().expr_ty(map_arg);
+    match map_closure_ty.kind() {
+        ty::Closure(_, _) | ty::FnDef(_, _) | ty::FnPtr(_) => {
+            let map_closure_sig = match map_closure_ty.kind() {
+                ty::Closure(_, substs) => substs.as_closure().sig(),
+                _ => map_closure_ty.fn_sig(cx.tcx),
+            };
+            let map_closure_return_ty = cx.tcx.erase_late_bound_regions(map_closure_sig.output());
+            is_type_diagnostic_item(cx, map_closure_return_ty, sym::Option)
         },
-        _ => {
-            return;
-        },
-    };
+        _ => false,
+    }
+}
 
-    let func_snippet = snippet(cx, map_arg.span, "..");
-    let hint = format!(".and_then({})", func_snippet);
-    let lint_info = format!("called `map(..).flatten()` on an `{}`", caller_type);
-    span_lint_and_sugg(
-        cx,
-        MAP_FLATTEN,
-        expr.span.with_lo(recv.span.hi()),
-        &lint_info,
-        "try using `and_then` instead",
-        hint,
-        Applicability::MachineApplicable,
-    );
+/// Split code suggestion into upper and bottom parts if its number of lines exceeds
+/// `MAX_SUGGESTION_HIGHLIGHT_LINES`, otherwise this function will just call `span_lint_and_sugg`
+/// with a combined `help` message.
+///
+/// This is useful when a suggested code snippet has multiple line thus causing possible useful
+/// information being chopped off at the end.
+fn span_lint_and_splitted_sugg(
```

Comment:
It would be good if you could move this to `clippy_utils::diagnostic` as @dswij suggested, with a note that only changes at the edge are shown and can be applied. Maybe `span_lint_and_help_for_edges` would be a better name that describes the goal of this function.

Reply:
should I name it as the one you suggested or `span_lint_and_sugg_for_edges`? It does fall back to `span_lint_and_sugg` after all

Reply:
Wait a minute, could a `span_lint_and_note` be used in this type of situation? 😆